slackscot.go

```diff
@@ -2,117 +2,133 @@ package slackscot
 
 import (
 	"fmt"
-	"github.com/alexandre-normand/slack"
 	"github.com/alexandre-normand/slackscot/config"
+	"github.com/nlopes/slack"
 	"log"
+	"os"
 	"regexp"
 )
 
-type Action struct {
-	// indicates whether the action should be omitted from the bot's help message.
-	// Defaults to false.
+// Slackscot represents what defines a Slack Mascot (mostly, a name and its plugins)
+type Slackscot struct {
+	name    string
+	plugins []Plugin
+	botEngine
+}
+
+// Plugin defines the lifecycle functions of a plugin: Init and Close
+type Plugin interface {
+	fmt.Stringer
+	Init(config.Configuration) (commands []ActionDefinition, listeners []ActionDefinition, err error)
+	Close()
+}
+
+// ActionDefinition represents how an action is triggered, published, used and described
+// along with defining the function defining its behavior
+type ActionDefinition struct {
+	// Indicates whether the action should be omitted from the help message
 	Hidden bool
 
-	// pattern to match in the message for the action's Answerer function to execute
+	// Pattern to match in the message for the action's Answerer function to execute
 	Regex *regexp.Regexp
 
-	// usage example
+	// Usage example
 	Usage string
 
-	// textual help description for the action
+	// Help description for the action
 	Description string
 
-	// function to execute if the Regex matches
+	// Function to execute if the Regex matches
 	Answerer ActionFunc
 }
 
-func (a Action) String() string {
+// String returns a friendly description of an ActionDefinition
+func (a ActionDefinition) String() string {
 	return fmt.Sprintf("`%s` - %s", a.Usage, a.Description)
 }
 
-type ActionFunc func(*slack.Message) string
+// ActionFunc is what gets executed when an ActionDefinition is triggered
+type ActionFunc func(ime *IncomingMessageEvent) string
 
-type Slackscot struct {
-	bundles []ExtentionBundle
-}
+// responseStrategy defines how a slack.OutgoingMessage is generated from a response
+type responseStrategy func(rtm *slack.RTM, ime *IncomingMessageEvent, response string) *slack.OutgoingMessage
 
-type ExtentionBundle interface {
-	fmt.Stringer
-	Init(config.Configuration) (commands []Action, listeners []Action, err error)
-	Close()
+// botEngine is the internal representation of a running bot
+type botEngine struct {
```

Comment:
This is bugging me a little bit. I don't really need those internal flattened lists of action definitions and this is getting clearer and clearer with other changes I'm working on. Don't expect this to be there much longer (but I'm probably not going to address until the next feature/PR).